Sources/SwiftDocC/Model/ParametersAndReturnValidator.swift

```diff
@@ -0,0 +1,610 @@
+        _ signatures: Signatures,
+        _ symbolKind: SymbolGraph.Symbol.Kind?,
+        hasDocumentedReturnValues: Bool
+    ) -> DocumentationDataVariants<ParametersSection> {
+        guard let parameters = parameters, !parameters.isEmpty else {
+            guard hasDocumentedReturnValues, Self.shouldAddObjectiveCErrorParameter(signatures, parameters) else {
+                // The symbol has no parameter documentation or return value documentation and none should be synthesized.
+                return DocumentationDataVariants(defaultVariantValue: nil)
+            }
+            // Since the symbol documented its return value synthesize an "error" parameter for it.
+            var variants = DocumentationDataVariants<ParametersSection>()
+            for trait in signatures.keys {
+                variants[trait] = ParametersSection(parameters: []) // Add an empty section so that this language doesn't fallback to the Objective-C content.
+            }
+            variants[.objectiveC] = ParametersSection(parameters: [Parameter(name: "error", contents: Self.objcErrorDescription)]) // This parameter is synthesized and doesn't have a source range.
+            return variants
+        }
+        
+        var variants = DocumentationDataVariants<ParametersSection>()
+        var allKnownParameterNames: Set<String> = []
+        var parameterNamesByExternalName: [String: Set<String>] = [:]
+        // Group the parameters by name to be able to diagnose parameters documented more than once.
+        let parametersByName = [String: [Parameter]](grouping: parameters, by: \.name)
+        
+        for (trait, signature) in signatures {
+            // Collect all language representations's parameter information from the function signatures.
+            let knownParameterNames = Set(signature.parameters.map(\.name))
+            allKnownParameterNames.formUnion(knownParameterNames)
+            
+            // Remove documented parameters that don't apply to this language representation's function signature.
+            var languageApplicableParametersByName = parametersByName.filter { knownParameterNames.contains($0.key) }
+            
+            // Add a missing error parameter documentation if needed.
+            if trait == .objectiveC, Self.shouldAddObjectiveCErrorParameter(signatures, parameters) {
+                languageApplicableParametersByName["error"] = [Parameter(name: "error", contents: Self.objcErrorDescription)] // This parameter is synthesized and doesn't have a source range.
+            }
+            
+            // Ensure that the parameters are displayed in the order that they need to be passed.
+            var sortedParameters: [Parameter] = []
+            sortedParameters.reserveCapacity(languageApplicableParametersByName.count)
+            for parameter in signature.parameters {
+                if let foundParameter = languageApplicableParametersByName[parameter.name]?.first {
```

Comment:
Can we be sure `first` will return the proper parameter? Does it matter? Is the call to `(grouping: by:)` above guaranteed to retain the original order?

Reply:
Yes, the order is [documented to be the same as the original order][1]. It mostly doesn't matter since documenting the same parameter results in a warning but it's nice that the behavior is stable across documentation builds even when there are warnings.

[1]: https://developer.apple.com/documentation/swift/dictionary/init(grouping:by:)

> The arrays in the “values” position of the new dictionary each contain at least one element, with the elements in the same order as the source sequence.